File: 09_OpenAPI/node/src/routers/spacecraftRouter.ts
import { Router } from 'express';

/** 
 *  @openapi 
 *  tags: 
 *    - name: Spacecrafts 
 *      description: Spacecraft API
 * 
 * 
 * components:
 *   schemas:
 *       SpacecraftId:
 *           description: The unique identifier of a spacecraft
 *           type: string
 *       Spacecraft:
 *           type: object
 *           required:
 *               - id
 *               - name
 *               - type
 *           properties:
 *               id:
 *                   $ref: "#/components/schemas/SpacecraftId"
 *               name:
 *                   type: string
 *               type:
 *                   type: string
 *                   enum:
 *                       - capsule
 *                       - probe
 *                       - satellite
 *                       - spaceplane
 *                       - station
 *               description:
 *                   type: string
 *       Error:
 *           type: object
 *           required:
 *               - message
 *           properties:
 *               message:
 *                   description: A human readable error message
 *                   type: string
 *   securitySchemes:
 *       ApiKey:
 *           type: apiKey
 *           in: header
 *           name: X-Api-Key
 * security:
 *   - ApiKey: []
*/

const router = Router();

const spacecrafts = [
    {
        id: 1,
        name: 'Apollo 11',
        type: 'capsule',
    },
    {
        id: 2,
        name: 'Apollo 13',
        type: 'capsule',
    },
];

/**
 * @openapi
 * /api/spacecrafts/{spacecraftId}:
 *    get:
 *      tags: [Spacecrafts]
 *      summary: Read a spacecraft
 *      responses:
 *          200:
 *              description: The spacecraft corresponding to the provided `spacecraftId`
 *              
 *          404:
 *              description: No spacecraft found for the provided `spacecraftId`
 *             
 *          500:
 *              description: Unexpected error
*/
router.get('/api/spacecrafts/:SpacecraftId', (req, res) => {
    const spacecraftId = parseInt(req.params.SpacecraftId);
    const spacecraft = spacecrafts.find((spacecraft) => spacecraft.id === spacecraftId);
    if (spacecraft) {
        res.json({ data: spacecraft });
    } else {
        res.status(404).json({ error: 'Spacecraft not found' });
    }
});

/**
 * @openapi
 * /api/spacecrafts:
 *    post:
 *     tags: [Spacecrafts]
 *     summary: Create a new spacecraft
 *     requestBody:
 *       description: The spacecraft to create
 *       required: true
 *       content:
 *         application/json:
 *           schema:
 *           $ref: '#/components/schemas/Spacecraft'
 *    responses:
 *      200:
 *        description: The spacecraft that was created
 *   
 *      500:
 *        description: Unexpected error
 */
router.post('/api/spacecrafts', (req, res) => {
    const spacecraft = req.body;
    spacecrafts.push(spacecraft);
    res.json({ data: spacecraft });
});



export default router;

File: 09_OpenAPI/python/routers/spacecraft_router.py
from fastapi import APIRouter, Query
from typing import Union
from pydantic import BaseModel

router = APIRouter()

spacecrafts = [
    {"id": 1, "Name": "Voyager 1", "Type": "Probe"},
    {"id": 2, "Name": "Voyager 2", "Type": "Probe"},
    {"id": 3, "Name": "Cassini", "Type": "Probe"},
    {"id": 4, "Name": "Pioneer 10", "Type": "Probe"},
    {"id": 5, "Name": "Pioneer 11", "Type": "Probe"},
    {"id": 6, "Name": "Viking 1", "Type": "Probe"},
]

class Spacecraft(BaseModel):
    id: int
    name: str
    type: str

@router.get("/spacecrafts/{spacecraft_id}")
def get_spacecraft(spacecraft_id: int, show_id: Union[str, None] = Query("Default", min_length=3, max_length=50)):
    for spacecraft in spacecrafts:
        if spacecraft["id"] == spacecraft_id:
            if show_id == "true":
                return spacecraft
            copy = spacecraft.copy()
            del copy["id"]
            return copy
    return {"Error": "Spacecraft not found"}

@router.post("/spacerafts")
def add_spacecraft(spacecraft: Spacecraft):
    spacecrafts.append(spacecraft)
    return spacecraft